sdk/src/types/block/core/mod.rs

```diff
@@ -0,0 +1,263 @@
+        highest_supported_version: u8,
+        protocol_parameters_hash: ProtocolParametersHash,
+    ) -> ValidationBlockBuilder {
+        ValidationBlockBuilder::new(strong_parents, highest_supported_version, protocol_parameters_hash)
+    }
+
+    /// Checks whether the block is a [`BasicBlock`].
+    pub fn is_basic(&self) -> bool {
+        matches!(self, Self::Basic(_))
+    }
+
+    /// Gets the block as an actual [`BasicBlock`].
+    /// NOTE: Will panic if the block is not a [`BasicBlock`].
+    pub fn as_basic(&self) -> &BasicBlock {
+        if let Self::Basic(block) = self {
+            block
+        } else {
+            panic!("invalid downcast of non-BasicBlock");
+        }
+    }
+
+    /// Checks whether the block is a [`ValidationBlock`].
+    pub fn is_validation(&self) -> bool {
+        matches!(self, Self::Validation(_))
+    }
+
+    /// Gets the block as an actual [`ValidationBlock`].
+    /// NOTE: Will panic if the block is not a [`ValidationBlock`].
+    pub fn as_validation(&self) -> &ValidationBlock {
+        if let Self::Validation(block) = self {
+            block
+        } else {
+            panic!("invalid downcast of non-ValidationBlock");
+        }
+    }
+
+    pub(crate) fn hash(&self) -> [u8; 32] {
+        Blake2b256::digest(self.pack_to_vec()).into()
+    }
+}
+
+impl Packable for Block {
+    type UnpackError = Error;
+    type UnpackVisitor = ProtocolParameters;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        match self {
+            Self::Basic(block) => {
+                BasicBlock::KIND.pack(packer)?;
+                block.pack(packer)
+            }
+            Self::Validation(block) => {
+                ValidationBlock::KIND.pack(packer)?;
+                block.pack(packer)
+            }
+        }?;
+
+        Ok(())
+    }
+
+    fn unpack<U: Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+        visitor: &Self::UnpackVisitor,
+    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        Ok(match u8::unpack::<_, VERIFY>(unpacker, &()).coerce()? {
+            BasicBlock::KIND => Self::from(BasicBlock::unpack::<_, VERIFY>(unpacker, visitor).coerce()?),
+            ValidationBlock::KIND => Self::from(ValidationBlock::unpack::<_, VERIFY>(unpacker, visitor).coerce()?),
+            k => return Err(Error::InvalidBlockKind(k)).map_err(UnpackError::Packable),
+        })
+    }
+}
+
+pub(crate) fn verify_parents(
+    strong_parents: &StrongParents,
+    weak_parents: &WeakParents,
+    shallow_like_parents: &ShallowLikeParents,
+) -> Result<(), Error> {
+    let (strong_parents, weak_parents, shallow_like_parents) = (
+        strong_parents.to_set(),
+        weak_parents.to_set(),
+        shallow_like_parents.to_set(),
+    );
+
+    if !weak_parents.is_disjoint(&strong_parents) || !weak_parents.is_disjoint(&shallow_like_parents) {
+        return Err(Error::NonDisjointParents);
+    }
```

Comment:
just to make sure: strong and shallow-like can be overlapping, right?

Reply:
Yes